### ISSUE TITLE
Non-contiguouos views

### ISSUE THREAD

**Issue by tsoernes** [bluss/ndarray#475]
Is there anything like the numpy equivalent of `X[np.where(X[:, i] < thresh)]; X[np.where(X[:, i] >= thresh)]` which will create two views of a 2D array, 1 with rows where column i is less than a threshold, and another view for the rest of the elements. 
```
/// Split `x` into two subviews: one with the rows where the value in the `feature_idx` column is below `threshold`,
//  and one where the value is equal or greater
fn split2<'a>(
    x: ArrayView2<'a, f64>,
    feature_idx: usize,
    threshold: f64,
) -> (ArrayView2<'a, f64>, ArrayView2<'a, f64>) {
    let idxs_lt: Vec<usize> = x.outer_iter()
        .enumerate()
        .filter(|(i, e)| e[[feature_idx]] < threshold)
        .map(|(i, e)| i)
        .collect();
    let idxs_gt: Vec<usize> = x.outer_iter()
        .enumerate()
        .filter(|(i, e)| e[[feature_idx]] >= threshold)
        .map(|(i, e)| i)
        .collect();
    let xl = x.reborrow().select(Axis(0), &idxs_lt);
    let xr = x.reborrow().select(Axis(0), &idxs_gt);
    // mismatched types (expected struct `ndarray::ViewRepr`, found struct `ndarray::OwnedRepr`) (rust-cargo)
    // expected type `ndarray::ArrayBase<ndarray::ViewRepr<&'a f64>, _>`
    // found type `ndarray::ArrayBase<ndarray::OwnedRepr<f64>, _>` (rust-cargo)
    // I don't want to copy the values in x
    (xl, xr)
}
```

**Comment by jturner314** [comment]
Note that NumPy copies the elements in the example you provided; it does not return a view. `np.where(X[:, i] < thresh` returns a single-element tuple containing an array of indices. Performing the `X[...]` slice operation with this tuple is an ["advanced indexing" operation, so it returns a copy of the data](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#advanced-indexing).

The Rust example you provided does the same thing as NumPy. (It determines the indices that meet the conditions, then copies the data corresponding to those indices into new arrays.)

In NumPy, if you really want to avoid copying the data, you could sort the array in-place, find the index where the threshold is between consecutive elements, and then split the array at that index (with two slices). Sorting the array in-place by a particular column  is tricky; [this is one way to do it](https://stackoverflow.com/a/2828371). Note that `a[a[:,i].argsort()]` copies the data, so that's not an option if you want to avoid the copy. An alternative approach would be to use [masked arrays](https://docs.scipy.org/doc/numpy/reference/maskedarray.generic.html).

With `ndarray`, if you really want to avoid copying the data, you could follow a similar approach. Again, sorting by a particular column is the tricky part. There isn't yet an equivalent for NumPy's masked arrays feature.

In conclusion, the simplest thing to do is to just copy the data.

To address the title of this issue specifically, it is possible to create non-contiguous views by [slicing with step sizes other than 1](https://docs.rs/ndarray/0.11.2/ndarray/macro.s.html). However, standard NumPy and `ndarray` arrays cannot represent views of an arbitrary sequence of indices; doing so would require a new type of array object. NumPy's "masked arrays" feature provides something like this.

**Comment by tsoernes** [comment]
Thank you for a very illuminating answer -- I was not aware that numpy behaved that way. Seeing that the behaviour I sought is not possible (for good reasons), I'll close this.
